File: app/src/contract/diagnostic.ts
// SPDX-License-Identifier: MIT
// Copyright (c) 2026 David Dunnock <dunnoda@gmail.com>
/**
 * A diagnostic, as the Rust core raises it.
 *
 * RUST COUNTERPART: `sv2_syntax::Diagnostic`, whose fields are private and read
 * through `code()`, `severity()`, `range()` and `message()`.
 *
 * SEVERITY IS CARRIED, NOT DERIVED HERE. On the Rust side severity comes from
 * the code — `DiagnosticCode::severity()` — so a caller cannot raise one code at
 * two severities. The webview could re-derive it from the same mapping, and
 * deliberately does not: that would be a second opinion about severity written
 * in a second language, which is the failure ADR-0013 DD-1 rejects for grammars
 * and which has the same shape here. One producer states it; this reads it.
 *
 * THE CODE IS VALIDATED BY SHAPE, NOT ENUMERATED. `DiagnosticCode` is
 * `#[non_exhaustive]` in Rust and today holds only the four `PARSE-*` codes;
 * `HIR-*`, `RES-*` and `ID-*` are planned. A closed enum here would make a
 * webview reject a diagnostic raised by a newer core — dropping the report that
 * something is wrong because the reason for it is unfamiliar, which inverts
 * ADR-0002: decorate by diagnostic state, never filter on it. So the schema
 * checks that a code is namespaced by its raising crate and admits it.
 *
 * A MESSAGE MAY CONTAIN MODEL TEXT. `sv2-cli` renders `unexpected \`x\``, where
 * `x` came out of the user's file. §9.3 forbids reporting model text to a log
 * channel, so `diagnostics/` must never forward one of these messages. It is a
 * value to render, not a value to log, and the thing to log beside it is the
 * code.
 *
 * NOT HERE YET: which file a diagnostic belongs to, and its line and column.
 * `sv2_syntax::Diagnostic` carries neither — it is per-parse, and `LineCol`
 * comes from `OffsetMap` over the text. Both belong to the workspace-level
 * query that returns diagnostics for more than one file, and arrive with it.
 */

import { z } from "zod";

import { type TextSpan, TextSpanSchema } from "./offset";

/**
 * How bad it is, mirroring `sv2_syntax::Severity::as_str`.
 *
 * Closed, unlike the code: a severity the webview does not know has no
 * rendering, so admitting one would buy nothing. Every code raised today is an
 * error — `sv2-syntax` has a test pinning that — and the other two exist
 * because ADR-0002 and ADR-0016 will need them.
 */
export const SEVERITIES = ["error", "warning", "info"] as const;

/** How bad it is. */
export type Severity = (typeof SEVERITIES)[number];

/** How bad it is. */
export const SeveritySchema: z.ZodType<Severity, unknown> = z.enum(SEVERITIES);

/**
 * A namespaced diagnostic code: the raising crate, then what happened.
 *
 * Anchored, uppercase, at least two segments — so `PARSE-EXPECTED` and
 * `PARSE-TOO-DEEPLY-NESTED` both pass and a bare `EXPECTED` does not. The
 * namespace is the part worth checking: it says which crate's vocabulary the
 * rest of the code belongs to, and it is what keeps `sv2-cli`'s process-level
 * `ErrorCode` from ever being mistaken for one of these.
 */
const DIAGNOSTIC_CODE = /^[A-Z]+(?:-[A-Z]+)+$/;

/** A namespaced diagnostic code, such as `PARSE-UNEXPECTED`. */
export const DiagnosticCodeSchema: z.ZodType<string, unknown> = z.string().regex(DIAGNOSTIC_CODE);

/**
 * The codes `sv2-syntax` raises today, from `DiagnosticCode::as_str`.
 *
 * A convenience for a caller that wants to say something specific about one of
 * them, and NOT the set the schema accepts — see the module comment. A code
 * absent from here is still a diagnostic and is still shown.
 */
export const PARSE_CODES = {
  expected: "PARSE-EXPECTED",
  unexpected: "PARSE-UNEXPECTED",
  unterminatedComment: "PARSE-UNTERMINATED-COMMENT",
  tooDeeplyNested: "PARSE-TOO-DEEPLY-NESTED",
} as const;

/** Something the core has to say about a span of a document. */
export type Diagnostic = Readonly<{
  code: string;
  severity: Severity;
  span: TextSpan;
  message: string;
}>;

/**
 * Something the core has to say about a span of a document.
 *
 * The message is not constrained beyond being a string. `sv2-syntax` documents
 * it as a lowercase phrase with no period and no position, but that is an
 * authoring convention on the raising side, and a schema that enforced it would
 * reject a legitimate diagnostic over punctuation — refusing to show the user
 * that their file is broken because the sentence describing it ended wrongly.
 */
export const DiagnosticSchema: z.ZodType<Diagnostic, unknown> = z.strictObject({
  code: DiagnosticCodeSchema,
  severity: SeveritySchema,
  span: TextSpanSchema,
  message: z.string(),
});

File: crates/sv2-syntax/src/diagnostic.rs
// SPDX-License-Identifier: MIT
// Copyright (c) 2026 David Dunnock <dunnoda@gmail.com>
//! What a tool says about a span of text.
//!
//! Defined here because `sv2-syntax` is the lowest crate that already owns
//! [`TextRange`], and every crate above it needs the type. One vocabulary, not one per
//! layer: codes are namespaced by the crate that raises them (`PARSE-*` here, `HIR-*`,
//! `RES-*` and `ID-*` above), and no crate defines its own enum for user-facing output.
//!
//! This is NOT `sv2-cli`'s `ErrorCode`'s job, and the two must not be merged.
//! That enum is a taxonomy of how the *process* failed — the file could
//! not be opened, the command line was not understood — and it maps one-to-one onto
//! exit statuses (STD-002-RS §3.2). A `Diagnostic` is about the *model*: it has a
//! position in a file, and a hundred of them still leave the process exit status at
//! "this file did not parse". A caller that could not tell those apart would report a
//! mistyped path as a grammar failure.
//!
//! A diagnostic carries a range so that it can be pointed at. ADR-0002 requires an
//! element with an unresolved reference to be *decorated* rather than filtered out, and
//! a decoration needs somewhere to go; an editor underlining a token needs the same
//! thing. Neither is expressible in a sentence of prose.

use text_size::TextRange;

/// How much a diagnostic matters.
///
/// [`Severity::Error`] is raised for text the language does not admit, and
/// [`Severity::Info`] for text admitted only by a recorded deviation from the
/// specification (`PARSE-DEVIATION`, ADR-0022), which the parser keeps apart from its
/// errors. `only_errors_and_deviations_are_raised` in the parser's tests holds that down.
/// [`Severity::Warning`] has no producer yet; ADR-0002 decorates rather than filters, and
/// ADR-0016 names an informational diagnostic of its own.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
#[non_exhaustive]
pub enum Severity {
    /// The text is not what the language admits. Gates a write; never gates a read.
    Error,
    /// Admitted, but probably not meant.
    Warning,
    /// Something the tool did that the author may want to know about.
    Info,
}

impl Severity {
    /// The severity's stable wire name, lowercase.
    #[must_use]
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Error => "error",
            Self::Warning => "warning",
            Self::Info => "info",
        }
    }
}

/// Every code a diagnostic raised by this crate can carry.
///
/// Closed, and `#[non_exhaustive]` so adding one is not a breaking change downstream
/// (STD-002-RS §2 rule 3). A code is added when something can produce it, never ahead
/// of that — the same rule `sv2-cli`'s `ErrorCode` follows. It cannot be linked from
/// here: `sv2-cli` is four layers up, and the dependency only points downward.
///
/// The `PARSE-` prefix is the namespace of the crate that raises it. A code from
/// `sv2-hir` will read `HIR-`, and the two will never collide.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
#[non_exhaustive]
pub enum DiagnosticCode {
    /// A production needed something here and found something else.
    Expected,
    /// Text that no production admits at this point.
    Unexpected,
    /// A `/* ... */` that the end of the file arrived in the middle of.
    UnterminatedComment,
    /// Input nested deeper than the parser will recurse (invariant 3).
    TooDeeplyNested,
    /// Text the parser admits only because of a recorded deviation from the
    /// specification's BNF, whose register entry the message names (ADR-0022).
    Deviation,
}

impl DiagnosticCode {
    /// The code's stable wire name, `SCREAMING-KEBAB-CASE` behind its namespace.
    ///
    /// Stable in the sense that matters: it is what a suppression comment, a test, or
    /// a downstream consumer names, so it may not be respelled once published.
    #[must_use]
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Expected => "PARSE-EXPECTED",
            Self::Unexpected => "PARSE-UNEXPECTED",
            Self::UnterminatedComment => "PARSE-UNTERMINATED-COMMENT",
            Self::TooDeeplyNested => "PARSE-TOO-DEEPLY-NESTED",
            Self::Deviation => "PARSE-DEVIATION",
        }
    }

    /// The severity a diagnostic with this code carries.
    ///
    /// A property of the code rather than of the site, so that two raisers of the same
    /// code cannot disagree about how much it matters.
    #[must_use]
    pub fn severity(self) -> Severity {
        match self {
            Self::Expected
            | Self::Unexpected
            | Self::UnterminatedComment
            | Self::TooDeeplyNested => Severity::Error,
            Self::Deviation => Severity::Info,
        }
    }
}

/// One thing a tool has to say about one span of text.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Diagnostic {
    code: DiagnosticCode,
    range: TextRange,
    message: String,
}

impl Diagnostic {
    /// A diagnostic at `range`, carrying `code` and a rendered `message`.
    ///
    /// The severity comes from the code, so a caller cannot raise the same code at two
    /// different severities.
    #[must_use]
    pub fn new(code: DiagnosticCode, range: TextRange, message: String) -> Self {
        Self {
            code,
            range,
            message,
        }
    }

    /// What kind of thing this is.
    #[must_use]
    pub fn code(&self) -> DiagnosticCode {
        self.code
    }

    /// How much it matters.
    #[must_use]
    pub fn severity(&self) -> Severity {
        self.code.severity()
    }

    /// The half-open byte range it is about.
    ///
    /// May be empty, and an empty range is not a defect: "expected `}`, found end of
    /// file" is about a position rather than about any bytes, and it still has to be
    /// pointed at.
    #[must_use]
    pub fn range(&self) -> TextRange {
        self.range
    }

    /// The rendered message, for a person.
    ///
    /// One lowercase phrase, no trailing period, and it never contains the position —
    /// the range is the position, and a caller renders it the way its output wants.
    #[must_use]
    pub fn message(&self) -> &str {
        &self.message
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Every code, so a variant added without a wire name or a severity fails here.
    const ALL: &[DiagnosticCode] = &[
        DiagnosticCode::Expected,
        DiagnosticCode::Unexpected,
        DiagnosticCode::UnterminatedComment,
        DiagnosticCode::TooDeeplyNested,
    ];

    #[test]
    fn every_code_is_namespaced_to_the_crate_that_raises_it() {
        for code in ALL {
            assert!(
                code.as_str().starts_with("PARSE-"),
                "{code:?} is raised by sv2-syntax and must say so"
            );
        }
    }

    #[test]
    fn no_two_codes_share_a_wire_name() {
        let mut seen: Vec<&str> = ALL.iter().map(|code| code.as_str()).collect();
        seen.sort_unstable();
        let count = seen.len();
        seen.dedup();
        assert_eq!(seen.len(), count, "two codes share a wire name");
    }

    #[test]
    fn severity_comes_from_the_code_and_not_from_the_site() {
        let range = TextRange::new(0.into(), 1.into());
        let one = Diagnostic::new(DiagnosticCode::Expected, range, "a".to_owned());
        let other = Diagnostic::new(DiagnosticCode::Expected, range, "b".to_owned());
        assert_eq!(one.severity(), other.severity());
        assert_eq!(one.severity(), Severity::Error);
    }

    #[test]
    fn an_empty_range_is_representable() {
        // "expected `}`, found end of file" is about a position, not about any bytes.
        let end = TextRange::empty(12.into());
        let diagnostic = Diagnostic::new(DiagnosticCode::Expected, end, "x".to_owned());
        assert!(diagnostic.range().is_empty());
        assert_eq!(diagnostic.range().start(), 12.into());
    }
}
